### PR TITLE
Add the saltext history, describe the module extraction process

### PR DESCRIPTION
**Pull request by max-arnold**

### CODE REVIEW

File: docs/topics/extraction.md [hunk 1]
```diff
@@ -1,5 +1,115 @@
...
+cd ..
+mkdir saltext-stalekey && cd saltext-stalekey
+git init
+copier copy --trust https://github.com/salt-extensions/salt-extension-copier ./
+```
+
+We need to remove some boilerplate files generated by Copier:
+
+```shell
+rm -rf tests/functional tests/integration tests/unit/engines/test_stalekey.py src/saltext/stalekey/engines/stalekey_mod.py
```
Comment (lkubb) [inline docs/topics/extraction.md]:
```suggestion
rm -f tests/**/test_*.py src/**/*_mod.py
```

suggestion: I would prefer if people would just delete the boilerplate intended to help write a new extension. `tests/functional/conftest.py` and `tests/integration/conftest.py` should be kept in case someone wants to add such tests in the future. They will also receive updates through Copier, unlike the above paths.

File: docs/topics/extraction.md [hunk 2]
```diff
@@ -1,5 +1,115 @@
...
+git remote rm repo-source
+git tag | xargs git tag -d
+```
+
+### 6. Clean up and testing
+
+Move the tests into the right location:
+
+```shell
+git mv tests/pytests/unit/engines/test_stalekey.py tests/unit/engines
```
Comment (lkubb) [inline docs/topics/extraction.md]: questions:

* Earlier you mention that `tests/unit/engines/test_stalekey.py` also exists. Doesn't this delete the non-Pytest tests?
* Do you know if non-Pytest tests work with the default saltext test suite? Never tried it, but assumed it would fail unless the whole translation layer was migrated from Salt core. Maybe we should add a note about conversion of tests and how to skip the whole test module to be converted at a later date.

File: docs/topics/extraction.md [hunk 3]
```diff
@@ -1,5 +1,115 @@
...
+Run the automatic fixups:
+
+```shell
+pip install git+https://github.com/saltstack/salt-rewrite
+SALTEXT_NAME=stalekey salt-rewrite -F fix_saltext .
+```
+
+```shell
+pip install -e ".[dev,tests,docs]"
+pre-commit run -a  # make sure it is happy
```
Comment (lkubb) [inline docs/topics/extraction.md]:
```suggestion
# Ensure you're inside your project's virtual environment from here on.
python -m venv venv --prompt saltext-stalekey
source venv/bin/activate
python -m pip install -e ".[dev,tests,docs]"
python -m pre_commit run -a   # make sure it is happy
python -m pre_commit install  # ensure the pre-commit hook runs when committing in the future
```

suggestion: I think we should point out to run this inside a venv.

File: docs/topics/extraction.md [hunk 4]
```diff
@@ -1,5 +1,115 @@
...
+
+Add the main authors to pyproject.toml:
+
+```shell
+vi pyproject.toml
+git add pyproject.toml
+git commit -m 'Add authors'
+```
+
+Try running the test suite locally until it passes, then commit and push it to run the full test suite on GitHub.
```
Comment (lkubb) [inline docs/topics/extraction.md]:
```suggestion
Try [running the test suite](run-tests-target) and [building the docs](build-docs-target) locally until both pass, then commit and push it to run the full test suite on GitHub.
```

suggestion: The docs are often an issue as well since the default configuration introduces a link checker.

File: docs/ref/concepts.md [hunk 5]
```diff
@@ -15,7 +15,32 @@ releases. On a downside, they need to be installed on each node separately.
...
+* In October 2023, a [project board](https://github.com/orgs/salt-extensions/projects/5/views/1) was created to track the migration process
+* After the Broadcom event, the push really came to shove, and it was said that there [won't be any deprecation period](https://www.youtube.com/watch?v=CubGR8rTy3Y&t=245s).
+* In January 2024, the [list of modules to be extracted](https://github.com/saltstack/great-module-migration) was opened for comments. Take your time to read it.
+* The modules are being migrated into the [salt-extensions](https://github.com/salt-extensions) community org.
+* In February 2024, the [great module purge PR](https://github.com/saltstack/salt/pull/65971) was created.
+* In April 2024, the great module purge PR was merged, making 3008 the target release.
+
+The main reasons for such a drastic change are:
+
+1. Reduce the amount of efforts required to maintain Salt.
```
Comment (lkubb) [inline docs/ref/concepts.md]:
```suggestion
1. Reduce the amount of effort required to maintain Salt.
```

File: docs/ref/concepts.md [hunk 6]
```diff
@@ -15,7 +15,32 @@ releases. On a downside, they need to be installed on each node separately.
...
+4. Release platform specific versions of Salt that are smaller and more efficient.
+5. Speed up PR merges and reviews through faster tests and a smaller core code base.
+
+The modules are split into three categories:
+
+1. Core modules - will be kept inside the main Salt codebase, and development will be tied to the Salt release cycle.
+2. Supported modules (extended core) - will be moved to their own repositories within the SaltStack Github organization where they can be maintained separately from the Salt codebase.
+3. Community modules - will be deprecated from the Salt Core codebase and community members will be able to continue independent maintainership if they are interested. The community modules can be hosted either in the community run Salt Extensions Github organization, or in individual or corporate source control systems.
+
+So, if you use any of the removed modules, consider becoming a maintainer or a contributor.
```
Comment (lkubb) [inline docs/ref/concepts.md]:
```suggestion
In conclusion, this migration represents a shift in how Salt is maintained and developed. It opens the door for users and organizations to have a direct impact on the tools they rely on. If you use any of the modules categorized as community modules, their future depends on people like you. By becoming a maintainer or contributor, you can ensure that the modules you depend on continue to thrive and evolve.
```

suggestion: I'd like to make this call-to-action more impactful. Full disclosure, I wrote a draft and ran it through an LLM, this is the (adjusted) result. Do you think it's over the top?

File: docs/topics/extraction.md [hunk 7]
```diff
@@ -1,5 +1,115 @@
...
+```
+
+### 2. Clone the Salt repo and analyze its history
+
+```shell
+git clone https://github.com/saltstack/salt --single-branch
+cd salt
+git filter-repo --analyze
+tree .git/filter-repo/analysis/
+rg stalekey .git/filter-repo/analysis/path-all-sizes.txt
```
Comment (lkubb) [inline docs/topics/extraction.md]: question: Is ripgrep required or does grep work just as fine? If the former, I think we should a Prerequisites heading and tell people to install it.

**Review by lkubb — CHANGES_REQUESTED** [review]
Wow, thanks a lot for taking the time to write this up!

I like it very much, just some questions/suggestions.

General question: Before publishing these docs, I intended to run most of them through an LLM to help improve clarity/conciseness. Do you approve of this in general? I would do it in a separate branch once the draft is finished and request your review.

File: docs/topics/extraction.md (hunk 2 above)
Comment (max-arnold) [reply docs/topics/extraction.md]: The non-pytest test no longer exists, it was removed even before the great purge PR. I included it in the filtered history for the sake of completeness (maybe someone will want to run git blame that far, to pre-pytest times).

> Do you know if non-Pytest tests work with the default saltext test suite?

I do not know this.

File: docs/ref/concepts.md (hunk 6 above)
Comment (max-arnold) [reply docs/ref/concepts.md]: No, you captured the spirit of it just right! I just had no time to write a more compelling paragraph :)

File: docs/topics/extraction.md (hunk 7 above)
Comment (max-arnold) [reply docs/topics/extraction.md]: Grep will work just fine, ripgrep is just what I had in my notes.

**Comment by max-arnold** [comment]
> General question: Before publishing these docs, I intended to run most of them through an LLM to help improve clarity/conciseness. Do you approve of this in general? I would do it in a separate branch once the draft is finished and request your review.

English is not my first language, so I'm totally fine with adding a little bit of polish using an LLM (as long as it isn't too wordy)

I'll incorporate your other suggestions a bit later

File: docs/topics/extraction.md (hunk 7 above)
Comment (lkubb) [reply docs/topics/extraction.md]: Let's `%s/rg/grep/g` then for minimalism's sake :)

**Review by lkubb — APPROVED** [review]
Perfect, thank you! <3

**Comment by lkubb** [comment]
@max-arnold Could you look over https://github.com/lkubb/saltext-copier/pull/2? It adjusts the steps a bit based on my replication attempt and adds a ref to my (quickly & dirtily written) new migration tool (https://github.com/salt-extensions/salt-extension-migrate) based on your guide.

Would appreciate your feedback on both. :)

Edit: If you want to stress-test a bit, try migrating `mysql`, `boto` and `proxmox`, for example. :)

**Comment by max-arnold** [comment]
I ran some tests, looks awesome! I expect some bugs though, so releasing it and getting some beta-testers would be the best option.

A couple of notes:

1. Is it possible/easy to display the output for long-running operations (git clone, analyze, filter)?
2. Can Copier wrap long questions (instead of truncating them) if they do not fit within the terminal width?
3. Can we make the Copier template use the rolling docs build process by default (on each push to the main branch)?
4. Some edge migration cases: modules that weren't removed (I know the tool detects those), utils that weren't removed https://github.com/saltstack/salt/tree/master/salt/utils/dockermod while the modules were
5. Maybe we can archive this repo https://github.com/salt-extensions/extension_migration ?

**Comment by lkubb** [comment]
> I ran some tests, looks awesome! I expect some bugs though, so releasing it and getting some beta-testers would be the best option.

Do you mean to PyPI and/or Discord? I intended for this tool to always be installed from the Git repo since it's very special-purpose and can be hotfixed easily this way (expecting bugs as well since all tests were manual). I'm not opposed to making it `pip(x)` installable though if you think that would be preferrable.

> 1. Is it possible/easy to display the output for long-running operations (git clone, analyze, filter)?

Yes, I initially thought it would make the output even more verbose without providing useful information, but maybe having feedback is better than waiting in the dark.

Adding it is just a `from plumbum import FG` and modifying the calls from (eg) `git("filter-repo", "--analyze")` to `git["filter-repo", "--analyze"] & FG` away.

> 2. Can Copier wrap long questions (instead of truncating them) if they do not fit within the terminal width?

That's a good question that has been bugging me as well. Copier uses `questionary` as well internally, I might need to check there.

Edit: Investigated this, this is caused by Copier adding a newline after the message that's passed to questionary, which borks the output.

> 3. Can we make the Copier template use the rolling docs build process by default (on each push to the main branch)?

That's already on `main` when `source_url` is inside the org, but the latest template tag points to before this commit. Wanted to get everything polished + the docs finished and then do a new release.

> 4. Some edge migration cases: modules that weren't removed (I know the tool detects those), utils that weren't removed [saltstack/salt@`master`/salt/utils/dockermod](https://github.com/saltstack/salt/tree/master/salt/utils/dockermod?rgh-link-date=2024-08-26T03%3A14%3A14Z) while the modules were

Yes, it detects all Python files that were not migrated. In essence, it just does a `ls **/*.py` and looks for non-empty output.

Some edge cases it accounts for as well:

**Renames:**
* renaming `tests/pytests`|`tests/support/pytest` to `tests/`|`tests/support` + there being non-pytest and pytest tests with the same name
* moving `salt/cloud/clouds` and `tests/{pytests/,}{unit,functional,integration}/cloud/clouds` one level up
* `doc` -> `docs`
* (not sure if necessary) moving `salt/client/ssh/wrapper` two levels up and renaming `tests/{pytests/,}integration/ssh` to `wrapper` (~~TODO I think the unit tests are in `tests/{pytests/,}unit/client/ssh/wrapper`~~)

**Copier defaults:**
* requiring test container support (grep `salt_factories.get_container`)

**Rewrites:**
* rewrites imports and `patch()` calls of most the above renames correctly (salt-rewrite does not)
* additionally rewrites all possible `__utils__` cases and warns about the rest (salt-rewrite fails to do this at all outside of Salt core):

1. If `__utils__` is not required: always rewrites the calls for both Salt core and migrated utils
2. `__utils__` used to call Saltext utils: rewrites calls partially, instructs to refactor the code again
3. `__utils__` used to call Salt core utils from Saltext utils: warns about failure and gives hints to resolve
4. `__utils__` used to call Salt core utils from other Saltext modules: Remarks that this works, but to consider submitting a PR to Salt core anyways

**Misc:**
* nudges to use `python3.10` in venv creation
* instructs to add `ref/utils/index` to `docs/ref/index.rst` if utils were migrated (this is neither done by copier nor the pre-commit hook)

Edge cases I know about:
* It fails when the Salt checkout currently is in the process of a rebase (e.g. caused by killing the tool at the wrong time - ~~TODO wrap this in try... finally~~)
* Afair an edge case during (non-)`pytest` renaming is when one of both does not exist anymore, but both existed at some point in the history together and were touched in the same commit (?) - `mysql` and commit `6abb43d2dfc362643989ed9a856ae38cf9d4c61e` touching `tests/integration/modules/test_mysql.py` + `tests/pytests/integration/modules/test_mysql.py` (does not exist anymore, moved to functional test) are an example which is fixed more by accident. I think `filter-repo` might fail if it was the other way around. If this happens, it can be worked around by passing `--avoid-collisions` though.
* It currently does not rewrite imports from other pytest test files. E.g. when a test does `from tests.pytests.unit.modules.test_foo import Bar`.

> 5. Maybe we can archive this repo [salt-extensions/extension_migration](https://github.com/salt-extensions/extension_migration) ?

💯 (Though we might need to migrate the issues somewhere else or just make the repository contents point to the migration guide here and keep it)

**Comment by max-arnold** [comment]
> Do you mean to PyPI and/or Discord?

Sorry, I wasn't clear enough. I meant just releasing (publishing) the salt-extension-copier guide and announcing it on Discord, so people can try porting some stuff using it.
